File: eg/sensor/sensor.c
/*
     ____ __     ____   ___    ____ __         (((((()
    | |_  \ \  /   ) ) | |  ) | |_  \ \  /  \(@)- /
    |_|__  \_\/  __)_) |_|_/  |_|__  \_\/   /(@)- \
                                               ((())))
 *//**
 *  \file  sensor.c
 *  \brief  ev3dev-c Sensors example.
 *  \author  Vitaly Kravtsov (in4lio@gmail.com)
 *  \copyright  See the LICENSE file.
 */

#include <string.h>
#include <stdio.h>
#include "ev3.h"
#include "ev3_port.h"
#include "ev3_sensor.h"

// WIN32 /////////////////////////////////////////
#ifdef __WIN32__

#include <windows.h>

// UNIX //////////////////////////////////////////
#else

#include <unistd.h>
#define Sleep( msec ) usleep(( msec ) * 1000 )

//////////////////////////////////////////////////
#endif

const char const *color[] = { "?", "BLACK", "BLUE", "GREEN", "YELLOW", "RED", "WHITE", "BROWN" };
#define COLOR_COUNT  (( int )( sizeof( color ) / sizeof( color[ 0 ])))

static bool _check_pressed( uint8_t sn )
{
	int val;

	if ( sn == SENSOR__NONE_ ) {
		return ( ev3_read_keys(( uint8_t *) &val ) && ( val & EV3_KEY_UP ));
	}
	return ( get_sensor_value( 0, sn, &val ) && ( val != 0 ));
}

int main( void )
{
	char s[ 256 ];
	int val;
	uint32_t n, i, ii;
	uint8_t sn_touch, sn_color, sn_ir;

	printf( "Waiting the EV3 brick online...\n" );
	if ( ev3_init() < 1 ) return ( 1 );

	printf( "*** ( EV3 ) Hello! ***\n" );
	ev3_sensor_init();

	printf( "Found sensors:\n" );
	for ( i = 0; i < DESC_LIMIT; i++ ) {
		if ( ev3_sensor[ i ].type_inx != SENSOR_TYPE__NONE_ ) {
			printf( "  type = %s\n", ev3_sensor_type( ev3_sensor[ i ].type_inx ));
			printf( "  port = %s\n", ev3_sensor_port_name( i, s ));
			if ( get_sensor_mode( i, s, sizeof( s ))) {
				printf( "  mode = %s\n", s );
			}
			if ( get_sensor_num_values( i, &n )) {
				for ( ii = 0; ii < n; ii++ ) {
					if ( get_sensor_value( ii, i, &val )) {
						printf( "  value%d = %d\n", ii, val );
					}
				}
			}
		}
	}
	if ( ev3_search_sensor( LEGO_EV3_IR, &sn_ir, 0 )) {
		printf( "IR sensor is found\n" );
	} else {
		printf( "IR sensor is NOT found\n" );
	}
	if ( ev3_search_sensor( LEGO_EV3_TOUCH, &sn_touch, 0 )) {
		printf( "TOUCH sensor is found, press BUTTON for EXIT...\n" );
	} else {
		printf( "TOUCH sensor is NOT found, press UP on the EV3 brick for EXIT...\n" );
	}
	if ( ev3_search_sensor( LEGO_EV3_COLOR, &sn_color, 0 )) {
		printf( "COLOR sensor is found, reading COLOR...\n" );
		set_sensor_mode( sn_color, "COL-COLOR" );
		for ( ; ; ) {
			if ( !get_sensor_value( 0, sn_color, &val ) || ( val < 0 ) || ( val >= COLOR_COUNT )) {
				val = 0;
			}
			printf( "\r(%s)", color[ val ]);
			fflush( stdout );
			if ( _check_pressed( sn_touch )) break;
			Sleep( 200 );
			printf( "\r        " );
			fflush( stdout );
			if ( _check_pressed( sn_touch )) break;
			Sleep( 200 );
		}
	} else {
		printf( "COLOR sensor is NOT found\n" );
		while ( !_check_pressed( sn_touch )) Sleep( 100 );
	}
	ev3_uninit();
	printf( "\n*** ( EV3 ) Bye! ***\n" );

	return ( 0 );
}

File: python/ev3dev/eg/sensor.py
# -*- coding: utf-8 -*-

from time import sleep
import sys
from ev3dev import *

color = [ '?', 'BLACK', 'BLUE', 'GREEN', 'YELLOW', 'RED', 'WHITE', 'BROWN' ]

def _check_pressed( sn ):
    if sn == SENSOR__NONE_:
        ok, val = ev3_read_keys()
        return ( ok and (( val & EV3_KEY_UP ) != 0 ))
    else:
        ok, val = get_sensor_value( 0, sn )
        return ( ok and ( val != 0 ))

if __name__ == '__main__':
    print 'Waiting the EV3 brick online...'
    if ev3_init() < 1: sys.exit( 1 )

    print '*** ( EV3 ) Hello! ***'
    ev3_sensor_init()

    print 'Found sensors:'
    for i in range( SENSOR_DESC__LIMIT_ ):
        type_inx = ev3_sensor_desc_type_inx( i )
        if type_inx != SENSOR_TYPE__NONE_:
            print '  type =', ev3_sensor_type( type_inx )
            print '  port =', ev3_sensor_port_name( i )
            ok, mode = get_sensor_mode( i, 256 )
            if ok:
                print '  mode =', mode
            ok, n = get_sensor_num_values( i )
            if ok:
                for ii in range( n ):
                    ok, val = get_sensor_value( ii, i )
                    if ok:
                        print '  value%d = %d' % ( ii, val )

    ok, sn_ir = ev3_search_sensor( LEGO_EV3_IR )
    if ok:
        print 'IR sensor is found'
    else:
        print 'IR sensor is NOT found'

    ok, sn_touch = ev3_search_sensor( LEGO_EV3_TOUCH )
    if ok:
        print 'TOUCH sensor is found, press BUTTON for EXIT...'
    else:
        print 'TOUCH sensor is NOT found, press UP on the EV3 brick for EXIT...'

    ok, sn_color = ev3_search_sensor( LEGO_EV3_COLOR )
    if ok:
        print 'COLOR sensor is found, reading COLOR...'
        set_sensor_mode( sn_color, 'COL-COLOR' )
        while ( 1 ):
            ok, val = get_sensor_value( 0, sn_color )
            if not ok or ( val < 0 ) or ( val >= len( color )):
                val = 0
            sys.stdout.write( '\r(%s)' % ( color[ val ]))
            sys.stdout.flush()
            if _check_pressed( sn_touch ):
                break
            sleep( 0.2 )
            sys.stdout.write( '\r        ' )
            sys.stdout.flush()
            if _check_pressed( sn_touch ):
                break
            sleep( 0.2 )
    else:
        print 'COLOR sensor is NOT found'
        while not _check_pressed( sn_touch ):
            sleep( 0.1 )

    ev3_uninit()
    print
    print '*** ( EV3 ) Bye! ***'
